src/app/clusters/tls-client-management-server/TlsClientManagementCluster.cpp

```diff
@@ -126,110 +145,135 @@ TlsClientManagementCluster::EncodeProvisionedEndpoints(EndpointId matterEndpoint
+    }
+    if (req.caid > kMaxRootCertId)
+    {
+        return Status::ConstraintError;
+    }
 
-    auto fabric = ctx.mCommandHandler.GetAccessingFabricIndex();
+    auto fabric     = commandHandler.GetAccessingFabricIndex();
+    auto endpointId = mPath.mEndpointId;
 
-    ReturnOnFailure(mCertificateTable.HasRootCertificateEntry(fabric, req.caid),
-                    ctx.mCommandHandler.AddStatus(
-                        ctx.mRequestPath, ClusterStatusCode::ClusterSpecificFailure(StatusCodeEnum::kRootCertificateNotFound)));
-    VerifyOrReturn(req.ccdid.IsNull() || mCertificateTable.HasClientCertificateEntry(fabric, req.ccdid.Value()) == CHIP_NO_ERROR,
-                   ctx.mCommandHandler.AddStatus(
-                       ctx.mRequestPath, ClusterStatusCode::ClusterSpecificFailure(StatusCodeEnum::kClientCertificateNotFound)));
+    if (mCertificateTable.HasRootCertificateEntry(fabric, req.caid) != CHIP_NO_ERROR)
+    {
+        return DataModel::ActionReturnStatus(ClusterStatusCode::ClusterSpecificFailure(StatusCodeEnum::kRootCertificateNotFound));
+    }
+    if (!req.ccdid.IsNull() && mCertificateTable.HasClientCertificateEntry(fabric, req.ccdid.Value()) != CHIP_NO_ERROR)
+    {
+        return DataModel::ActionReturnStatus(ClusterStatusCode::ClusterSpecificFailure(StatusCodeEnum::kClientCertificateNotFound));
+    }
 
     Commands::ProvisionEndpointResponse::Type response;
-    auto status = mDelegate.ProvisionEndpoint(ctx.mRequestPath.mEndpointId, fabric, req, response.endpointID);
+    auto status = mDelegate.ProvisionEndpoint(endpointId, fabric, req, response.endpointID);
 
     if (status.IsSuccess())
     {
-        ctx.mCommandHandler.AddResponse(ctx.mRequestPath, response);
-
-        MatterReportingAttributeChangeCallback(ctx.mRequestPath.mEndpointId, TlsClientManagement::Id,
-                                               TlsClientManagement::Attributes::ProvisionedEndpoints::Id);
-    }
-    else
-    {
-        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, status);
+        ConcreteCommandPath responsePath(mPath.mEndpointId, TlsClientManagement::Id, Commands::ProvisionEndpoint::Id);
+        commandHandler.AddResponse(responsePath, response);
+        NotifyAttributeChanged(TlsClientManagement::Attributes::ProvisionedEndpoints::Id);
+        return std::nullopt;
     }
+
+    return DataModel::ActionReturnStatus(status);
 }
 
-void TlsClientManagementCluster::HandleFindEndpoint(HandlerContext & ctx, const Commands::FindEndpoint::DecodableType & req)
+std::optional<DataModel::ActionReturnStatus>
+TlsClientManagementCluster::HandleFindEndpoint(CommandHandler & commandHandler, const Commands::FindEndpoint::DecodableType & req)
 {
     ChipLogDetail(Zcl, "TlsClientManagement: FindEndpoint");
 
-    VerifyOrReturn(req.endpointID <= kMaxTlsEndpointId, ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::ConstraintError));
+    if (req.endpointID > kMaxTlsEndpointId)
+    {
+        return Status::ConstraintError;
+    }
```

Comment:
```suggestion
    VerifyOrReturnError(req.endpointID <= kMaxTlsEndpointId, Status::ConstraintError);
```

is an often used pattern.